libqtile/window.py

```diff
@@ -93,6 +93,13 @@
 _NET_WM_STATE_ADD = 1
 _NET_WM_STATE_TOGGLE = 2
 
+# z layers
+DESKTOP_LAYER = -9999
+BELOW_LAYER = -99
+NORMAL_LAYER = 0
+ABOVE_LAYER = 99
+FULLSCREEN_LAYER = 9999
```

Comment:
Any reason you picked these values rather than others? I feel it's weird that we jump from 99 to 9999.

Also, what would you think of handling fullscreen windows outside of the Z axis? i.e. whatever the Z-index a window have, if there's a fullscreen window, it will be drawn upon it. Then we wouldn't have to need an upper bound, and thus a limit in the number of windows we can order.

That may be the opposite of what I said previously, and I'm sorry if that's the case! But I think we should ensure we only ever have one fullscreen window, and hide all other windows, so it's quite a special case.

Reply:
There is not much reasoning behind the specific numbers.

I wanted to handle all of these cases to adhere to https://specifications.freedesktop.org/wm-spec/1.3/ar01s07.html#STACKINGORDER . There might be cases where multiple fullscreen applications exist (or not?), and even cases where they have floating child windows (which get a "+1") or where the user *wants* windows above a fullscreen window. I will think about this for a bit, but currently I think this is a good approach.

Reply:
> There might be cases where multiple fullscreen applications exist (or not?)

What's the point?

> and even cases where they have floating child windows (which get a "+1") or where the user wants windows above a fullscreen window.

Well, right, you might want your notifications over a movie, for example. But we should not always try to handle all the edge cases, and rather have a strong opinion when it makes sense. I think it would be simpler and quite safe to take the decision that a fullscreen window is above everything, always. That would adhere to the said specification and I don't think anyone would ever complain about this, especially since we have the Max layout that could be used to circumvent this. What do you think?

Reply:
> What's the point?
Well, a max layout with a hidden bar has the same effect, so there isn't too much of a point.

Initially I actually started by splitting the windows three ways - "normal", "floating" and "fullscreen". The 9999 wouldn't be required in that case, since fullscreen windows will stay on top of others anyway. It did make things a bit easier, but I could go back to that behaviour.

Reply:
I removed the 9999-value; fullscreen-windows now have an initial z-value of 0, but they are kept above everything else.

Reply:
I was just testing a few things and opened up a VirtualBox-window in the process. Put it into fullscreen, opened a settings-window and... I couldn't see it, since the fullscreen window stays on top of everything. I think that's not enough of an edge case to ignore.
In addition I installed Tint2 to check out #1145 (I assume that was about floating windows covering the bar), and that isn't fixed yet.

So I actually think we should include the 9999, plus maybe change the behaviour of `group.assure_correct_layer` to allow for any window to be on top of any other window, if the z-value demands it. That would mean no more priorization just by type/state of the window alone.

Reply:
> I was just testing a few things and opened up a VirtualBox-window in the process. Put it into fullscreen, opened a settings-window and... I couldn't see it, since the fullscreen window stays on top of everything. I think that's not enough of an edge case to ignore.

Right. But then what Z value would these window have?

Reply:
How about 1000000 (10^6) for fullscreen? Any child-window will get a plus one, respectively. I am using it like that right now, and it works well. Do you see the problem with a fixed value in general or with the specific value of 9999 (as it was before)?

We could go back and remove the section that assigns hard values for specific windows (above, below, etc. - X11 specs), with every window just starting with a 0. But then there would be some more logic needed to handle child-windows or they could be hidden by their parents again. And handling those X11-flags was probably a good idea to begin with.

Reply:
In addition to notifications and child windows, people also use utilities such as dmenu/rofi that create windows that we would expect to be drawn at the very top. This might be a non-issue though, as tools such as those are usually untouched by the WM.

Do we need to handle fullscreen windows differently? What are the issues with the current behaviour? Maybe all they would need is to bring_to_front when fullscreen is enabled, on the occasion when a window that is covered is in focus and fullscreened.

Reply:
I use dmenu and I have never had any issues. Can't say the same about general popups (#1893).

The only thing that separates a fullscreen window from the common window in this PR is a somewhat higher starting z-value. I wouldn't consider that "handling them differently". And if they had the same z-value as other windows, this PR would be kind of obsolete, wouldn't it?

Reply:
Fair enough :+1: